File: week3/Selecting the Optimal k_mer Size.py
# python3

"""
Coursera/Genome Assembly Programming Challenge/Week 3/Problem 2:
Selecting the Optimal k-mer Size
Grading: 'Good job! (Max time used: 0.11/15.00, max memory used: 11403264/536870912.)'
Author: Andrii Shostatskyi
Email: shostatskyi.andrii@gmail.com
Respect Coursera Honor Code
Copyright © 2018. All rights reserved
"""

import sys
from enum import Enum


class Result(Enum):
    NoCycle = 1
    OneCycle = 2
    MultipleCycles = 3


class NoOptimalKFound(Exception):
    pass


def check_Euler_cycle(de_Bruijn_graph):
    for k, v in de_Bruijn_graph.items():
        if not v: 
            return Result.NoCycle
        if len(v) > 1: 
            return Result.MultipleCycles

    return Result.OneCycle


def break_reads_into_kmers(reads, k):
    break_read = lambda read: [ read[j:j + k] for j in range(len(read) - k + 1) ]
    return [ kmer for read in reads for kmer in break_read(read) ]


def build_de_Bruijn_graph(kmers):
    graph = {}
    for kmer in kmers:
        left, right = kmer[:-1], kmer[1:]
        if left != right:
            graph.setdefault(left, set()).add(right)
            graph.setdefault(right, set())

    return graph


def binary_search_optimal_k(reads, l, r):
    while r >= l:
        mid = int(l + (r - l) / 2)
        result = check_Euler_cycle(build_de_Bruijn_graph(break_reads_into_kmers(reads, mid)))

        if result == Result.NoCycle:
            r = mid - 1
            continue
        elif Result == Result.MultipleCycles:
            l = mid + 1
            continue
        else:
            return mid

    raise NoOptimalKFound()
 

if __name__ == "__main__":
    k, reads = 100, sys.stdin.read().split()
    print(binary_search_optimal_k(reads, 0, k))

File: week3/Selecting the Optimal k_mer Size.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <set>
#include <stack>

enum class result
{
    no_cycle,
    one_cycle,
    multiple_cycles
};

result check_Euler_cycle(const std::map<std::string, std::set<std::string>> graph)
{
    for(auto & kv : graph)
    {
        if(kv.second.empty()) return result::no_cycle;
        if(kv.second.size() > 1) return result::multiple_cycles;
    }

    return result::one_cycle;
}

std::map<std::string, std::set<std::string>>
construct_DeBruijn_graph(const std::vector<std::string> & reads, const int k)
{
    std::map<std::string, std::set<std::string>> graph;

    for(const auto & read : reads) 
    {
        for(size_t i = 0; i + k < read.size(); ++i)
        {
            graph[read.substr(i, k-1)].emplace(read.substr(i + 2, k - 1));
            if(i + k + 1 < read.size()) graph[read.substr(i + 2, k - 1)];
        }
    }

    return graph;
}

std::vector<std::string> read_input()
{
    std::ios::sync_with_stdio(false);
    std::cin.tie(NULL);

    std::vector<std::string> reads;
    std::string s;

    while(std::cin >> s) 
        reads.emplace_back(std::move(s));

    return reads;
}

int binary_search_k(const std::vector<std::string> reads, int l, int r) 
{ 
    while(r >= l) 
    { 
        int mid = l + (r - l) / 2; 

        result res =  check_Euler_cycle(construct_DeBruijn_graph(reads, mid));
        switch(res)
        {
            case result::one_cycle: 
                return mid;
            case result::no_cycle:
                r = mid - 1;
                continue; 
            case result::multiple_cycles:
                l = mid + 1;
                continue;
        }
    } 

    throw std::logic_error{"optimal k not found"};
} 

int main ()
{
    const int mer_size = 100;    
    std::cout << binary_search_k(read_input(), 0, mer_size) << std::endl;

    return 0;
}